File: day03.py
import re
import csv
from collections import Counter


CLAIM_PATTERN = re.compile(r'^#(?P<id>\d+) @ (?P<x>\d+),(?P<y>\d+): (?P<width>\d+)x(?P<height>\d+)')
COLLISION_CHAR = 'X'
FABRIC_SIZE = 1000


def claim_parser(claim):
    match = re.match(CLAIM_PATTERN, claim)
    return tuple(int(x) for x in match.groups())


with open('day03-input.txt', 'r') as input_file:
    claims = [claim_parser(row.strip()) for row in input_file]

# Init fabric
fabric = [[''] * FABRIC_SIZE for _ in range(FABRIC_SIZE)]

collisions = 0

for claim in claims:
    id, x, y, width, height = claim

    for i in range(width):
        for j in range(height):
            current_x = x + i
            current_y = y + j
            existing_value = fabric[current_y][current_x]
            insert_value = id
            if existing_value:
                if existing_value in [id, COLLISION_CHAR]:
                    continue
                # Fresh collision
                collisions += 1
                insert_value = COLLISION_CHAR
            fabric[current_y][current_x] = insert_value

# pt 1
print(collisions)

# Lazy style 😬
# Just do the same thing again to find the one that doesn't overlap.
for claim in claims:
    id, x, y, width, height = claim
    collided = False

    for i in range(width):
        for j in range(height):
            existing_value = fabric[y + j][x + i]
            if existing_value and existing_value != id:
                collided = True
                continue
    if not collided:
        # pt 2
        print(id)
        # Only 1 claim doesn't collide.
        break

# Visuallize the fabric in a spreadsheet
with open('day03-fabric.csv', 'w') as csv_file:
    writer = csv.writer(csv_file)
    writer.writerows(fabric)

File: day03.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strconv"
)

type claim struct {
	id, x, y, width, height int
}

func parseClaim(str string) claim {
	// fmt.Println(str)
	r := regexp.MustCompile(`#(\d+) @ (\d+),(\d+): (\d+)x(\d+)`)
	var g []int
	for _, group := range r.FindStringSubmatch(str)[1:] {
		asInt, _ := strconv.Atoi(group)
		g = append(g, asInt)
	}
	return claim{g[0], g[1], g[2], g[3], g[4]}
}

func main() {
	file, err := os.Open("day03-input.txt")
	if err != nil {
		panic(err)
	}
	var claims []claim
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		claims = append(claims, parseClaim(scanner.Text()))
	}

	const (
		fabricSize    = 1000
		collisionChar = "X"
	)
	var fabric [fabricSize][fabricSize]string
	collisions := 0
	for _, claim := range claims {
		for i := 0; i < claim.width; i++ {
			for j := 0; j < claim.height; j++ {
				currentX := claim.x + i
				currentY := claim.y + j

				existingValue := fabric[currentY][currentX]
				insertValue := strconv.Itoa(claim.id)

				if existingValue != "" {
					if existingValue == strconv.Itoa(claim.id) || existingValue == collisionChar {
						continue
					}
					// Fresh collision
					collisions++
					insertValue = collisionChar
				}
				fabric[currentY][currentX] = insertValue
			}
		}
	}
	// pt 1
	fmt.Println(collisions)

	// Lazy style 😬
	// Just do the same thing again to find the one that doesn't overlap.
	for _, claim := range claims {
		collided := false

		for i := 0; i < claim.width; i++ {
			for j := 0; j < claim.height; j++ {
				existingValue := fabric[claim.y+j][claim.x+i]
				if existingValue != "" && existingValue != strconv.Itoa(claim.id) {
					collided = true
					// wasted effort here, should go to next claim
					break
				}
			}
		}
		if !collided {
			// pt 2
			fmt.Println(claim.id)
			break
		}
	}
}
